src/Microsoft.AspNet.Routing/Template/TemplateMatcher.cs

```diff
@@ -40,126 +40,175 @@ public class TemplateMatcher
-                    {
-                        return null;
-                    }
+                    Debug.Assert(!routeSegment.IsSimple);
+
+                    // Don't attempt to validate a complex segment at this point other than being non-emtpy,
+                    // do it in the second pass.
                 }
             }
 
-            for (var i = requestSegments.Count; i < Template.Segments.Count; i++)
+            for (; i < Template.Segments.Count; i++)
             {
                 // We've matched the request path so far, but still have remaining route segments. These need
                 // to be all single-part parameter segments with default values or else they won't match.
-                var routeSegment = Template.Segments[i];
-                if (routeSegment.Parts.Count > 1)
+                var routeSegment = Template.GetSegment(i);
+                Debug.Assert(routeSegment != null);
+
+                if (!routeSegment.IsSimple)
                 {
-                    // If it has more than one part it must contain literals, so it can't match.
+                    // If the segment is a complex segment, it MUST contain literals, and we've parsed the full
+                    // path so far, so it can't match.
                     return null;
                 }
 
                 var part = routeSegment.Parts[0];
                 if (part.IsLiteral)
                 {
+                    // If the segment is a simple literal - which need the URL to provide a value, so we don't match.
                     return null;
                 }
 
-                Debug.Assert(part.IsParameter);
+                if (part.IsCatchAll)
+                {
+                    // Nothing to validate for a catch-all - it can match any string, including the empty string.
+                    //
+                    // Also, a catch-all has to be the last part, so we're done.
+                    break;
+                }
 
-                // It's ok for a catch-all to produce a null value
-                object defaultValue;
-                if (Defaults.TryGetValue(part.Name, out defaultValue) || part.IsCatchAll)
+                // If we get here, this is a simple segment with a parameter. We need it to be optional, or for the
+                // defaults to have a value.
+                Debug.Assert(routeSegment.IsSimple && part.IsParameter);
+                if (!Defaults.ContainsKey(part.Name) && !part.IsOptional)
                 {
-                    values.Add(part.Name, defaultValue);
+                    // There's no default for this (non-optional) parameter so it can't match.
+                    return null;
                 }
-                else if (part.IsOptional)
+            }
+
+            // At this point we've very likely got a match, so start capturing values for real.
+            var values = new RouteValueDictionary();
```

Comment:
I actually like the ugliness of this. I think it's good. If there were two methods, they have to be maintained as one unit, and there's no strong contract (types) making them be the same.